### DISCUSSION TITLE
RoadRunner `v2025.1` ideas discussion

### DISCUSSION THREAD

**Discussion by rustatian** [roadrunner-server/roadrunner#1960] [Ideas]
Discussions for the ideas from: https://github.com/roadrunner-server/roadrunner/issues/1959

**Comment by Kaspiman** [comment]
Let's add a few ultimate features to kill FrankenPHP

**Reply by rustatian** [reply]
FrankenPHP raised a good idea in the PHP community. I respect the guys behind that project. I don't agree with the concepts of merging Go and C via CGO and forcing to use Caddy (and this is totally ok), but competition is a great thing. We can learn something from them, while they can learn our approaches and apply them for the good PHP future. So, no need to kill anyone 😃

**Reply by Kaspiman** [reply]
I expected something like: `Agree.`

**Comment by rustatian** [comment]
@Kaspiman https://github.com/roadrunner-server/roadrunner/issues/1959#issuecomment-2198292302
1. Need to think about that. Generally, I agree that from `lower to higher` value might be a better representation of the priority.
2. Yes, agree.
3. Also agree, but to support both commands while deprecating `reset`. For 1 year, showing a deprecation method.
4. Agree.
5. Already exists.

**Reply by Kaspiman** [reply]
5. Look at my Grafana with rr_jobs* metrics - there is no queue size gauge.
![image](https://github.com/roadrunner-server/roadrunner/assets/10486803/c7e526b8-a179-4052-b287-376d2e22bafc)


HTTP-pluging metrics looks like this:
![image](https://github.com/roadrunner-server/roadrunner/assets/10486803/5f376323-f4f2-4696-9346-7e8b4a40d4f6)

**Reply by rustatian** [reply]
Yeah, it exists, but not exposed. Generally, we should maintain the same mechanism for queue reporting across all plugins. The workers pool already supports the [`QueueSize`](https://github.com/roadrunner-server/pool/blob/master/pool/static_pool/pool.go#L363) method.

**Reply by Kaspiman** [reply]
i remembered some old task about requeue tag [Add requeue tag to failed task metric of Jobs plugin](https://github.com/roadrunner-server/roadrunner/issues/1566 )

**Comment by Kaspiman** [comment]
How about upgrade RR logo? It doesn't look very memorable or special right now

**Reply by rustatian** [reply]
Thanks for the proposal, but logo is fine, no need to update it.

**Comment by Kaspiman** [comment]
Hello again! I found out that [BoltDB (KV driver)](https://github.com/boltdb/bolt) has been out of development for about six years now (latest commit).

Should we find a new alternative for this? For example [Badger](https://github.com/dgraph-io/badger)

**Reply by rustatian** [reply]
Hey @Kaspiman 👋 
I'm not using this driver.

**Reply by Kaspiman** [reply]
Oh, i see. RR use [fork bbolt](https://github.com/etcd-io/bbolt), not original repo.

**Comment by Kaspiman** [comment]
Hello. There is an idea to change the mechanism of delayed messages for AMQP-driver.

Now we declare dead-letter queue for every [delayed task](https://github.com/roadrunner-server/amqp/blob/master/amqpjobs/driver.go#L674).  The result is a large number of queues with different names.

We have more than 100 queues on the project, many of them with pending tasks. A large number of queues affects the [cardinality of metrics](https://docs.victoriametrics.com/faq/#what-is-high-cardinality) and can lead to disruption of metrics rendering. Also we try to send messages to queues through a storage mediator/accumulator. This mediator helps us not to lose tasks in case of network connectivity failure.

 What if we stopped creating new temporary queues, leaving only one per queue? For example `delayed_queue_name`. Let's change [dlx-behavior per queue](https://www.rabbitmq.com/docs/ttl#per-queue-message-ttl) to [dlx per message](https://www.rabbitmq.com/docs/ttl#per-message-ttl-in-publishers).

When creating a pipeline, you can immediately create a delayed queue. What do you think?

**Reply by rustatian** [reply]
Hey @Kaspiman 👋 
The case is, IIRC, that every message for the `dlx` queue has to be unique, because `x-dead-letter-exchange` and `x-dead-letter-routing-key` are `Queue` properties, not messages. While TTL's yes, might be set per-message, it is used to automatically resend the message to the `routing-key` and `exchange` specified in the dead-letter-queue.

I thought about that in the past, and it might be valid to use only 1 dead-letter-queue bound to the pipeline's exchange and queue. And when we have a delayed message or dead-lettered message, we can use message properties to retry, yeah.

**Reply by Kaspiman** [reply]
Of course we need to create delayed queue for every pipeline separatelly). You'll have to create twice as many queues, but it's worth it. Recently I tried to implement such a mechanism by adding temporary/virtual RR-pipelines, but it didn’t work out.

**Reply by Kaspiman** [reply]
Is this change possible in a minor version? Of course, changing the names of queues looks like a compatibility breakdown. However, those queues are temporary and are automatically deleted over time. Introducing new prefixed queues shouldn't break current projects too much.

**Reply by rustatian** [reply]
This won't break existing projects because dlx mechanism automatically sends the data to the configured exchange and routing key. Even if we update this mechanism to use only 1 DLX queue created at runtime, other dlx queues, previously created by the old RR, will anyway send their pending messages. But this is not a bugfix change, and so, will wait for at least 2024.3 (fall).

**Reply by Kaspiman** [reply]
Hi. I tested my idea with one delayed queue. It just doesn't work. Reason: messages in RabbitMQ are dying from the head of the queue. If there are messages with a big delay in the head - they will prevent messages with a small delay in the tail of the queue from dying. This is why everyone uses temporary DLX-queues with a delay in the name instead one DLX-queue.

**Reply by rustatian** [reply]
Yeah, I guess I remember why I did DLX exactly the way I did them 😄

**Comment by Kaspiman** [comment]
Hello! Let's change AMQP default options to improve durability:
* [queue durable](https://docs.roadrunner.dev/docs/queues-and-jobs/amqp#durable) `false -> true`
* [exchange durable](https://docs.roadrunner.dev/docs/queues-and-jobs/amqp#exchange-durable) `false -> true`

Also let's change similar params in `roadrunner-php/jobs` package:
* [queue durable](https://github.com/roadrunner-php/jobs/blob/9842316c03a4a21030ac78cd8829245a849d60d0/src/Queue/AMQPCreateInfo.php#L15) `false -> true`
* [exchange durable](https://github.com/roadrunner-php/jobs/blob/9842316c03a4a21030ac78cd8829245a849d60d0/src/Queue/AMQPCreateInfo.php#L20) `false -> true`

**Reply by rustatian** [reply]
Hey, I'm not sure that we should touch configurable parameters for no significant reason.

**Reply by Kaspiman** [reply]
The reason is very good.

Classic queues will survive restarts and will not lose messages. Modern quorum queues have this mode by default, and it’s great. It makes sense to promote classic queues to be durable. It is difficult to imagine a situation where non-durable queues are needed.

* https://www.rabbitmq.com/docs/queues#durability
* https://www.rabbitmq.com/blog/2023/03/02/quorum-queues-migration#non-durable-queues

```
Non-durable queues
Non-durable queues will be deleted on a node/cluster boot. Having extra durability guarantees that mirroring provides is a bit pointless.

Non-durable queues concept is also going away in the future releases: the only option for ephemeral queues will be exclusive queues. This affects only durability of queue definitions, messages can still be marked transient.

For such queues a decision have to be made one way or another: is this queue content important enough to get availability guarantees of quorum queues, or it's better to downgrade it to a classic (but durable) queue.
```

**Reply by rustatian** [reply]
I understand the reasoning, but not everyone needs this. 
The default value for `bool` is false, everybody knows that and expects, that the `bool` value for the configuration would be false by default.

**Comment by Kaspiman** [comment]
Hello! Let's take a look at [pool.allocate_timeout](https://docs.roadrunner.dev/docs/error-codes/allocate-timeout). There are two different operations with one timeout:
1. allocate workers at RR start
2. allocate request/job at free worker (essentially a timeout for a request/job to be in a queue)

Maybe we should separate timeout by operations?

There is another idea: to collect the internal queue settings in a separate section:
```
pool:
    internal_queue:
        max_size: 100
        timeout: 10s
````

**Reply by rustatian** [reply]
Hey @Kaspiman 👋 
Thanks for the proposal.
There is no need for general users to know about internal implementation details. Allocate timeout is already a known option and no need to change it.

**Reply by Kaspiman** [reply]
I was trying to figure out this parameter right now and it turns out I need to change it. These two operations need to be set to different timeouts. Creating a worker and taking a request from the queue are completely different operations, why do they have the same config? Also, the word ["allocate"](https://docs.roadrunner.dev/docs/error-codes/allocate-timeout) does not accurately describe the taking of a request from a queue to a worker.

**Reply by rustatian** [reply]
Thanks for worrying about that. Sorry, but this option won't be changed. 
Reasoning:
That is straightforward to change everything in the configuration to correspond to the every internal piece. The question is: Who would use that? My realistic answer to that: possibly a few devs + we have to introduce a BC to config, update version, support 2 configuration versions, etc.
'Allocate' word meaning is: distribute (resources) for a particular purpose. Looking at our use-case, we 'Allocate' (aka distribute) workers to use them to handle a request. So the word is used correctly from my POV.

**Comment by ryan053857** [comment]
I think there's a rr can be based on [static-php-cli](https://github.com/crazywhalecc/static-php-cli) and PHP running environment of independent binary package, similar frankenphp official like [that](https://frankenphp.dev/docs/static/)

**Comment by rustatian** [comment]
Crazy idea, but why not. The current problem is that RR supports only plugins which are written in Golang. What if we go beyond and think about WIT (Wasm) and wasm-based-modules. Having a good interface written in WIT (https://component-model.bytecodealliance.org/design/wit.html) we basically would be able to connect plugins written in any language which supports WASM and WIT. The good example is how `Zed` editor handles extensions: [blog](https://zed.dev/blog/zed-decoded-extensions). In the same way, we'd be able to connect any extension to RR.
The issue is, that in Go, unlike Rust, WIT is not fully supported yet.
Having such extensions, we'd be able to have middleware in PHP for example, or better gRPC streaming integration an so on.

**Comment by fugikzl** [comment]
About metrics for jobs, we have in our project gauge metrics for tasks in queue, that helps us to monitor system status and it's very helpful. I vote for it with both hands
![image](https://github.com/user-attachments/assets/b9e8472c-badd-4792-862e-3fc7c89d0a95)

The code is simple, and executes every 10 seconds as service
```php
$collectorType = CollectorType::Gauge;
foreach (((array)$rpc->call('jobs.Stat', null))['stats'] as $queue) {
    $metrics->set($collectorType->value . '_' . $queue['queue'], $queue['active'] ?? 0);
}
```

**Comment by maximal** [comment]
@rustatian, maybe, RoadRunner could consider Server-Sent Events support in the nearest future?

* https://developer.mozilla.org/docs/Web/API/Server-sent_events
* https://mercure.rocks/

**Reply by rustatian** [reply]
Why not WebSockets via Centrifugo? https://centrifugal.dev
